[READY TO MERGE]: PR for video events support

HritwikSinghal opened the pull request:
# Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
Fixes #1446 


## Status
**Ready to merge**


## Type of change

<!-- Please delete options that are not relevant. -->


- New feature (non-breaking change which adds functionality)
- This change requires a documentation update


# How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. -->

- Completely tested

# Checklist:

- [x] I have performed a self-review of my own code
- [x] New and existing unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have made corresponding changes to the documentation (if applicable)

FabioRosado commented:
Yeah sure, anything that you might think it could be useful for other developers when working with video events 👍

subodhawasthi commented:
Here is the Requirements Document we've made,
[Requirements Document.docx](https://github.com/opsdroid/opsdroid/files/5529516/Requirements.Document.docx)

FabioRosado commented:
That's an interesting document and the project seems pretty well structured. I've seen that you have added a few constraints to the video formats. I hope you will not be too hard on this, the reason for that is that I'm assuming the service (like telegram) will accept any format and will do some sort of conversion, so if you upload a `mov` file, everything should work fine. 

Hope this makes sense. I'm eager to see what you come up with to fix this issue 👍

subodhawasthi commented:
Yeah sure, we'll include it.

HritwikSinghal commented:
Hey @FabioRosado , we are planning to use ffmpeg to return the video details. Using ffmpeg will give the devs all the info that the video contained. And this way opsdroid can support all video formats that are currently available in ffmpeg (which are almost all major formats available) 
Here is an example of output that we are planning on returning


            "codec_name": "h264",
            "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
            "profile": "High",
            "codec_type": "video",
            "codec_time_base": "1/60",
            "codec_tag_string": "avc1",
            "codec_tag": "0x31637661",
            "width": 1920,
            "height": 1080,
            "coded_width": 1920,
            "coded_height": 1088,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p",
            "level": 40,
            "color_range": "tv",
            "color_space": "bt709",
            "color_transfer": "bt709",
            "color_primaries": "bt709",
            "chroma_location": "left",
            "refs": 1,
            "is_avc": "true",
            "nal_length_size": "4",
            "r_frame_rate": "30/1",
            "avg_frame_rate": "30/1",
            "time_base": "1/15360",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 1057280,
            "duration": "68.833333",
            "bit_rate": "2314520",
            "bits_per_raw_sample": "8",
            "nb_frames": "2065",
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0


However, the the machine on which opsdroid is running must have ffmpeg installed. this is not a problem for linux machines since ffmpeg comes pre-installed on them but windows devs must install it separately. Should we proceed? if yes where should we put the updated requirements?

FabioRosado commented:
Hello, I understand that you are working through this problem, could I ask you to keep your local branch up to date by using `git pull upstream` instead of updating this WIP PR? Thank you

HritwikSinghal commented:
> Hello, I understand that you are working through this problem, could I ask you to keep your local branch up to date by using `git pull upstream` instead of updating this WIP PR? Thank you

Yes, i will take care of it. Hey can you help us as to why these many tests are failing on the commit that i pushed?

FabioRosado commented:
The e2e tests don't seem to fail because of this PR so perhaps it's not related to this. The linting is failing because of a few things. I'm happy to give a proper review once the PR is finished 👍

HritwikSinghal commented:
> The e2e tests don't seem to fail because of this PR so perhaps it's not related to this. The linting is failing because of a few things. I'm happy to give a proper review once the PR is finished +1

Hey, the PR is complete from my side, please review it and let me know what is left.

jacobtomlinson commented:
> Oh and we need to add a package to be installed on systems, but there are no 'requirements.txt' in the folder. so where should we add it.

You can find the requirements listed in `setup.cfg`.

jacobtomlinson commented on opsdroid/events.py:
```diff
@@ -413,6 +417,37 @@ def __init__(self, name=None, params=None, *args, **kwargs):  # noqa: D107
+    async def save_to_disk(self, fbytes):
+        """ Save the bytes to disk as video."""
+
+        temp_dir = tempfile.gettempdir()
+        file_name = "opsdroid_vid_" + await str(random.randint(0, 100000000))
+        file_path = os.path.join(temp_dir, file_name)
+
+        # Checks and deletes the output file
+        # You cant have a existing file or it will through an error
+        if os.path.isfile(file_path):
+            os.remove(file_path)
+
+        # opens the file which is accessable as 'out_file'
+        async with open(file_path, "wb") as out_file:  # open for [w]riting as [b]inary
+            await out_file.write(fbytes)
```
Have a look at [`NamedTemporaryFile`](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) it exists for exactly this use case.

jacobtomlinson commented on opsdroid/events.py:
```diff
@@ -413,6 +417,37 @@ def __init__(self, name=None, params=None, *args, **kwargs):  # noqa: D107
+        return file_path
+
+    async def get_video_properties(self):
+        """ Get the video properties like codec, resolution."""
+        """ Returns Video properties saved in Dictionary"""
```
```suggestion
        """Get the video properties like codec, resolution.
        
        Returns Video properties saved in Dictionary.
        
        """
```

HritwikSinghal commented:
> > Oh and we need to add a package to be installed on systems, but there are no 'requirements.txt' in the folder. so where should we add it.
> 
> You can find the requirements listed in `setup.cfg`.

Thanks, This PR is completed from our side. We also added the binary representation of the video that previous PR was implementing. All the tests should pass and all components are being tested. The CI on github is failing however.

jacobtomlinson commented:
> All the tests should pass and all components are being tested. The CI on github is failing however.

Your tests are not passing, the CI is showing this to you.

```
   >       self.assertEqual(await str(event.get_properties()), self.props)
  E       TypeError: object str can't be used in 'await' expression
```

Please fix this line.

HritwikSinghal commented:
> > All the tests should pass and all components are being tested. The CI on github is failing however.
> 
> Your tests are not passing, the CI is showing this to you.
> 
> ```
>   >       self.assertEqual(await str(event.get_properties()), self.props)
>   E       TypeError: object str can't be used in 'await' expression
> ```
> 
> Please fix this line.

Hey jacob, now all tests will pass. The CI for ubuntu and Windows is failing since "ffmpeg" is not installed in these environments by default. So what should i do.

jacobtomlinson commented:
> The CI for ubuntu and Windows is failing since "ffmpeg" is not installed in these environments by default. So what should i do. ?

If ffmpeg is a dependency for your changes then we need to handle that. 

For the CI we should add ffmpeg. But we also need to fail gracefully if the user does not have ffmpeg installed.

HritwikSinghal commented:
> > The CI for ubuntu and Windows is failing since "ffmpeg" is not installed in these environments by default. So what should i do. ?
> 
> If ffmpeg is a dependency for your changes then we need to handle that.
> 
> For the CI we should add ffmpeg. But we also need to fail gracefully if the user does not have ffmpeg installed.

ffmpeg is a dependency for Video class only. If the developer (on who's machine opsdroid is installed) never interacts with video class ( more specifically the function "get_properties()" ) , then the program wont fail. 
However if ffmpeg is not installed in the OS (and/or not added to PATH ), then the Video class will definitely throw error. 

for ubuntu:
```
sudo add-apt-repository universe -y
sudo apt update -y
sudo apt install ffmpeg -y
```

for Arch based:
```
sudo pacman -S ffmpeg
```

for Fedora, RHEL, CentOS Based:
```
sudo dnf install ffmpeg -y
sudo dnf -y install ffmpeg-devel
```

NOTE: These commands may change in future or may become invalid, so user has to make sure 'ffmpeg' is installed AND added to PATH in OS.

jacobtomlinson commented:
> However if ffmpeg is not installed in the OS (and/or not added to PATH ), then the Video class will definitely throw error.

Yes and you need to catch that error and show something useful to the user. Show a warning saying that video events are not supported unless ffmpeg is installed.

HritwikSinghal commented:
```
        try:
            "Something"
        except RuntimeError as error:
            _LOGGER.warning(_("Video events are not supported unless ffmpeg is installed."))
            print("Video events are not supported unless ffmpeg is installed.")
```

Would this be enough to notify user?

jacobtomlinson commented:
Just the log, not the print.

HritwikSinghal commented:
> Just the log, not the print.

should i change in the CI config for installing ffmpeg in ubuntu or you would do it? if yes, then where should i do it. And the tests will fail for windows anyway since installing ffmpeg and adding it to path is manual process in windows.

FabioRosado commented on opsdroid/events.py:
```diff
@@ -413,6 +415,44 @@ def __init__(self, name=None, params=None, *args, **kwargs):  # noqa: D107
+
+    async def get_properties(self):
+        """Get the video properties like codec, resolution.
+        Returns Video properties saved in a Dictionary"""
```
```suggestion
        Returns Video properties saved in a Dictionary
        
        """
```

FabioRosado commented on opsdroid/events.py:
```diff
@@ -413,6 +415,44 @@ def __init__(self, name=None, params=None, *args, **kwargs):  # noqa: D107
+        fbytes = await self.get_file_bytes()
+
+        # create a temp_vid file to store the bytes
+        temp_vid = tempfile.NamedTemporaryFile(prefix="opsdroid_vid_")
```
Does this temp file get removed after we have done what we needed with it?

HritwikSinghal replied:
> Hello,
> 
> I've added a few suggestions and questions on the code. I need to think about what would be best on the test side. If getting ffmpeg installed or maybe use a mock there thinking
> 
> Also, since ffmpeg is a requirement to make this work, can you update the documentation to mention this? We should add it there so users know that this is expected otherwise people will only know they need ffmpeg after the first failure.

Yeah sure i will make changes to the documentation! can you help me where should i do that?

And for ubuntu you can add the 3 lines (mentioned above) to install ffmpeg in tt, the main problem is for windows since its hard to install ffmpeg using cmd line there.

HritwikSinghal replied:
it does after returning the value, but i have added a line to manually deleting that temp file after returning the value in next commit.

FabioRosado reviewed (changes requested):
Hello, 

I've added a few suggestions and questions on the code. I need to think about what would be best on the test side. If getting ffmpeg installed or maybe use a mock there 🤔 

Also, since ffmpeg is a requirement to make this work, can you update the documentation to mention this? We should add it there so users know that this is expected otherwise people will only know they need  ffmpeg after the first failure.

HritwikSinghal commented:
> Hello,
> 
> I've added a few suggestions and questions on the code. I need to think about what would be best on the test side. If getting ffmpeg installed or maybe use a mock there thinking
> 
> Also, since ffmpeg is a requirement to make this work, can you update the documentation to mention this? We should add it there so users know that this is expected otherwise people will only know they need ffmpeg after the first failure.

Yeah sure i will make changes to the documentation! can you help me where should i do that?

And for ubuntu you can add the 3 lines (mentioned above) to install ffmpeg in tt, the main problem is for windows since its hard to install ffmpeg using cmd line there.

HritwikSinghal commented:
I updated github CI file and fixed lint issues in latest commit. i skipped tests on windows so you can think about another solution to it. Meanwhile all tests are passing now. Also please tell me where to update documentation for telling users to install ffmpeg.

jacobtomlinson reviewed (commented):
This is really great. Thank you so much with sticking through the review process, it has been a long one.

If you do not have `ffmpeg` installed locally do the tests fail? If so we should skip them. 

You've added tests here in the old unittest style. We are migrating everything to pytest style which would make this easier. 

I'm tempted to just get this merged and address this in a future PR. Perhaps you would be interested in working on it?

HritwikSinghal commented:
> This is really great. Thank you so much with sticking through the review process, it has been a long one.
> 
> If you do not have `ffmpeg` installed locally do the tests fail? If so we should skip them.
> 
> You've added tests here in the old unittest style. We are migrating everything to pytest style which would make this easier.
> 
> I'm tempted to just get this merged and address this in a future PR. Perhaps you would be interested in working on it?

The tests would fail if ffmpeg is not installed. Skipping the tests may not be a good idea since i found a bug in windows due to the fact that tests were failing on windows only. We can add the ffmpeg to the requirements and update the docs to notify the users. 

As for the pytest style, If you can open an issue regarding the migration of tests to pytest style (for the whole project) , i may change these in future PR (i would have to learn pytest and then fix the tests)

jacobtomlinson commented:
> Skipping the tests may not be a good idea since i found a bug in windows due to the fact that tests were failing on windows only.

We should definitely run these in CI, but skip locally when developing.

> As for the pytest style, If you can open an issue regarding the migration of tests to pytest style (for the whole project) , i may change these in future PR (i would have to learn pytest and then fix the tests)

Already exists #1502

HritwikSinghal commented:
> We should definitely run these in CI, but skip locally when developing.

ok, i will make changes in next PR. is there any environment variable set on CI so that i can add that condition?

> Already exists #1502
Ok, i will try to migrate those tests if i get time.

jacobtomlinson commented:
> is there any environment variable set on CI so that i can add that condition?

In pytest you can use [`pytest.skip`](https://docs.pytest.org/en/latest/skipping.html#skipping-test-functions) to skip over tests. And there is a `skipif` decorator. So you could do something like this.

```python
import pytest
import shutil

@pytest.mark.skipif(shutil.which("ffmpeg") is None, reason="Test requires ffmpeg")
def test_which_requires_ffmpeg():
    # Do something with ffmpeg
```
